File: backend/tour-service/src/main/java/com/project/tour/service/tour/operation/ActivityCruiseTourAssignmentService.java
package com.project.tour.service.tour.operation;

import com.project.tour.dto.tour.operation.ActivityCruiseTourAssignmentRequest;
import com.project.tour.dto.tour.operation.ActivityCruiseTourAssignmentResponse;
import com.project.tour.exception.AppException;
import com.project.tour.mapper.tour.ActivityCruiseTourAssignmentMapper;
import com.project.tour.model.ActivityCruiseTour;
import com.project.tour.model.CruiseArea;
import com.project.tour.model.Tour;
import com.project.tour.model.enums.onboard.ActivityCruiseTourStatus;
import com.project.tour.model.enums.tour.TourStatusTrip;
import com.project.tour.repository.cruise.CruiseAreaRepository;
import com.project.tour.repository.tour.ActivityCruiseTourAssignmentRepository;
import com.project.tour.repository.tour.TourRepository;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class ActivityCruiseTourAssignmentService {

        private final ActivityCruiseTourAssignmentRepository assignmentRepository;
        private final TourRepository tourRepository;
        private final CruiseAreaRepository cruiseAreaRepository;

        public ActivityCruiseTourAssignmentService(
                        ActivityCruiseTourAssignmentRepository assignmentRepository,
                        TourRepository tourRepository,
                        CruiseAreaRepository cruiseAreaRepository) {

                this.assignmentRepository = assignmentRepository;
                this.tourRepository = tourRepository;
                this.cruiseAreaRepository = cruiseAreaRepository;
        }

        /**
         * Operation phân công một khu vực cho Tour.
         *
         * Chỉ tạo assignment.
         *
         * activityCruise = null
         * startTime = null
         * endTime = null
         * maxPassengers = null
         * price = null
         *
         * status = WAITING_CONFIG
         */
        public ActivityCruiseTourAssignmentResponse assign(
                        ActivityCruiseTourAssignmentRequest request) {

                Tour tour = tourRepository
                                .findById(request.tourId())
                                .orElseThrow(() -> new AppException(
                                                "Tour not found",
                                                HttpStatus.NOT_FOUND));

                // NẾU TOUR CẦN Ở TRẠNG THÁI PENDING HOẶC APPROVED THÌ CÓ THỂ BỎ BỚT HOẶC ĐIỀU
                // CHỈNH CHECK STATUS Ở ĐÂY
                // Ví dụ: Cho phép phân công khi Tour đang chờ duyệt hoặc đã duyệt

                CruiseArea cruiseArea = cruiseAreaRepository
                                .findById(request.cruiseAreaId())
                                .orElseThrow(() -> new AppException(
                                                "Cruise area not found",
                                                HttpStatus.NOT_FOUND));

                if (cruiseArea.getStatus() == null) {
                        throw new AppException(
                                        "Cruise area status is invalid",
                                        HttpStatus.BAD_REQUEST);
                }

                if (cruiseArea.getCruiseDeck() == null) {
                        throw new AppException(
                                        "Cruise area is not assigned to a deck",
                                        HttpStatus.BAD_REQUEST);
                }

                if (tour.getCruise() == null) {
                        throw new AppException(
                                        "Tour has not been assigned to a cruise",
                                        HttpStatus.BAD_REQUEST);
                }

                if (!tour.getCruise().getId()
                                .equals(cruiseArea.getCruiseDeck().getCruise().getId())) {

                        throw new AppException(
                                        "Cruise area does not belong to the cruise assigned to this tour",
                                        HttpStatus.BAD_REQUEST);
                }

                // Chống trùng lặp bản ghi
                return assignmentRepository.findByTourIdAndCruiseAreaId(request.tourId(), request.cruiseAreaId())
                                .map(ActivityCruiseTourAssignmentMapper::toResponse)
                                .orElseGet(() -> {
                                        ActivityCruiseTour assignment = new ActivityCruiseTour();
                                        assignment.setTour(tour);
                                        assignment.setCruiseArea(cruiseArea);
                                        assignment.setActivityCruise(null);
                                        assignment.setStartTime(null);
                                        assignment.setEndTime(null);
                                        assignment.setMaxPassengers(null);
                                        assignment.setPrice(null);
                                        assignment.setStatus(ActivityCruiseTourStatus.WAITING_CONFIG);

                                        ActivityCruiseTour saved = assignmentRepository.save(assignment);
                                        return ActivityCruiseTourAssignmentMapper.toResponse(saved);
                                });
        }

        /**
         * Lấy toàn bộ phân công của một Tour.
         */
        @Transactional(readOnly = true)
        public List<ActivityCruiseTourAssignmentResponse> getByTour(
                        UUID tourId) {

                if (!tourRepository.existsById(tourId)) {
                        throw new AppException(
                                        "Tour not found",
                                        HttpStatus.NOT_FOUND);
                }

                return assignmentRepository
                                .findAllByTourIdOrderByCreatedAtAsc(tourId)
                                .stream()
                                .map(ActivityCruiseTourAssignmentMapper::toResponse)
                                .toList();
        }

        /**
         * Xóa phân công khi chưa được Onboard cấu hình.
         */
        @Transactional
        public void deleteByTourAndArea(UUID tourId, UUID cruiseAreaId) {
                ActivityCruiseTour assignment = assignmentRepository
                                .findByTourIdAndCruiseAreaId(tourId, cruiseAreaId)
                                .orElseThrow(() -> new AppException("Assignment not found", HttpStatus.NOT_FOUND));

                if (assignment.getStatus() != ActivityCruiseTourStatus.WAITING_CONFIG) {
                        throw new AppException(
                                        "Cannot delete an assignment that has already been configured",
                                        HttpStatus.BAD_REQUEST);
                }

                assignmentRepository.delete(assignment);
        }
}

File: frontend-web/src/modules/operation/services/activityCruiseTourAssignmentService.js
// src/modules/operation/services/activityCruiseTourAssignmentService.js

import api from "../../../api/axios";

const BASE_URL = "/operation/activity-cruise-tour-assignment";
const CONFIGURED_BASE_URL = "/operation/activity-cruise-tours";

const activityCruiseTourAssignmentService = {
  /**
   * =========================================================
   * API PHÂN CÔNG CŨ
   * =========================================================
   */

  /**
   * GET /api/operation/activity-cruise-tour-assignment/tour/{tourId}
   *
   * Lấy danh sách các khu vực đã phân công hoạt động cho Tour.
   */
  getByTour: async (tourId) => {
    const response = await api.get(`${BASE_URL}/tour/${tourId}`);

    return response.data;
  },

  /**
   * POST /api/operation/activity-cruise-tour-assignment
   *
   * Phân công 1 khu vực Hoạt động cho Tour.
   * payload = { tourId, cruiseAreaId }
   */
  assign: async (payload) => {
    const response = await api.post(BASE_URL, payload, {
      headers: {
        "Content-Type": "application/json",
      },
    });

    return response.data;
  },

  /**
   * DELETE
   * /api/operation/activity-cruise-tour-assignment/tour/{tourId}/area/{cruiseAreaId}
   *
   * Xóa phân công khu vực hoạt động.
   */
  delete: async (tourId, cruiseAreaId) => {
    const response = await api.delete(
      `${BASE_URL}/tour/${tourId}/area/${cruiseAreaId}`,
    );

    return response.data;
  },

  /**
   * =========================================================
   * API NHẬN CẤU HÌNH ĐÃ HOÀN THÀNH
   * =========================================================
   */

  /**
   * GET /api/operation/activity-cruise-tours
   *
   * Lấy tất cả ActivityCruiseTour đã được cấu hình.
   */
  getAllConfigured: async () => {
    const response = await api.get(CONFIGURED_BASE_URL);

    return response.data?.data ?? response.data;
  },

  /**
   * GET /api/operation/activity-cruise-tours/tour/{tourId}
   *
   * Lấy các ActivityCruiseTour đã cấu hình của một Tour.
   */
  getConfiguredByTour: async (tourId) => {
    const response = await api.get(`${CONFIGURED_BASE_URL}/tour/${tourId}`);

    return response.data?.data ?? response.data;
  },
};

export default activityCruiseTourAssignmentService;
